Jetty 12 simplify Retainable.canRetain usage

gregw opened the pull request:
I believe that `canRetain` is no longer needed. The only place it is used for real logic is in `AsyncContent`, which can use it's own chunk class to pass on how to succeed the callback.

gregw commented:
@sbordet @lorban the main change in this PR is in `AsyncContent`, everything else is just a mechanical removal of `canRetain`.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java:
```diff
@@ -502,12 +541,9 @@ static Chunk from(ByteBuffer byteBuffer, boolean last, Consumer<ByteBuffer> rele
          */
         static Chunk asChunk(ByteBuffer byteBuffer, boolean last, Retainable retainable)
         {
-            if (!retainable.canRetain())
```
This is the most important spot where `canRetain` is used. We must not allow linking the lifecycle of a Chunk that potentially contains a pooled buffer to the lifecycle of non-retaining retainables like EOF or EMPTY.

While this check is a nice to have extra safeguard and the current code would work without it, it was added because that error is way too easy to make and hard to track down.

lorban reviewed (changes requested):
I've added a comment with a quick summary of our discussion around `Content.asChunk` and `canRetain` for future reference.

lorban reviewed (approved):
LGTM

gregw committed:
Cleanup non-retainable `Retainable`s

`Retainable`s that return false from `canRetain` now are noops if `retain` is called, which allows for a simpler calling convention.
AsyncContent has also been reworked to allocate less and be clearer in its use of `canRetain`.

gregw commented:
@sbordet @lorban rebased against HEAD of jetty-12.0.x after merge of #9073
